flow/layers/display_list_layer.cc

```diff
@@ -89,16 +89,16 @@ void DisplayListLayer::Preroll(PrerollContext* context,
 
   DisplayList* disp_list = display_list();
 
-  if (auto* cache = context->raster_cache) {
+  if (context->raster_cache && context->raster_cache->PreCheckWillCache(
+                                   disp_list, is_complex_, will_change_)) {
     TRACE_EVENT0("flutter", "DisplayListLayer::RasterCache (Preroll)");
 
     SkMatrix ctm = matrix;
     ctm.preTranslate(offset_.x(), offset_.y());
 #ifndef SUPPORT_FRACTIONAL_TRANSLATION
     ctm = RasterCache::GetIntegralTransCTM(ctm);
 #endif
```

Comment:
It seems like this patch could be simplified quite a bit by just moving this into `RasterCache::Prepare` and passing the offset instead of the ctm.

I'm also curious how much time this actually saves, but I'd be supportive of a simplification of the API like that even if it doesn't actually save many cycles :)

Reply:
I'm not very sure about this. Offset is applied based on matrix set by `ScopedFrame`, and finally used by canvas when rasterize, if we abandon  matrix in here seems will lose a lot information.

And yes it won't save a lot speed in release mode, but may still has some benefit when a lot picture layer in one screen.

Reply:
Sorry, I think I mixed something up. We shouldn't change that to `offset`, but perhaps we can change the API to expect the untransformed matrix rather than an already transformed one. 

```cpp
  bool Prepare(GrDirectContext* context,
               SkPicture* picture,
               const SkMatrix& transformation_matrix,
               SkColorSpace* dst_color_space,
               bool is_complex,
               bool will_change);
  bool Prepare(GrDirectContext* context,
               DisplayList* display_list,
               const SkMatrix& transformation_matrix,
               SkColorSpace* dst_color_space,
               bool is_complex,
               bool will_change);
```

to

```
  bool Prepare(GrDirectContext* context,
               SkPicture* picture,
               const SkMatrix& untransformed_matrix,
               SkColorSpace* dst_color_space,
               bool is_complex,
               bool will_change);
  bool Prepare(GrDirectContext* context,
               DisplayList* display_list,
               const SkMatrix& transformation_matrix,
               SkColorSpace* dst_color_space,
               bool is_complex,
               bool will_change);
```

And then we could just guard the common matrix transformation logic behind whether we rasterize the picture, and avoid the need for an additional function call. WDYT?

Reply:
I agree, matrix translate can before rasterize since decomposition and raster cache key won't affect by translate

Reply:
So now that this has moved to `Prepare`, do we still need `PrecheckWillCache`? It seems like that logic can just live in `Prepare` now before the matrix decomposition, and we can just call `prepare` like before.